pennylane/templates/layers/gate_fabric.py

```diff
@@ -0,0 +1,222 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""
+Contains the quantum number preserving GateFabric template.
+"""
+# pylint: disable-msg=too-many-branches,too-many-arguments,protected-access
+import numpy as np
+import pennylane as qml
+from pennylane.operation import Operation, AnyWires
+
+
+class GateFabric(Operation):
+    r"""Implements a local, expressive, and quantum-number-preserving ansatz using VQE circuit fabrics
+    proposed by Anselmetti *et al.* in `arXiv:2104.05692 <https://arxiv.org/abs/2104.05695>`_.
+
+    This template prepares the :math:`N` qubits trial state by applying :math:`D` layers of gate fabric block
+    :math:`\hat{U}_{GF}(\vec{\theta},\vec{\phi})` to the Hartree-Fock state in the Jordan-Wigner basis
+
+    .. math::
+
+        \vert \Psi(\vec{\theta},\vec{\phi})\rangle =
+        \hat{U}_{GF}^{(D)}(\vec{\theta}_{D},\vec{\phi}_{D}) \ldots
+        \hat{U}_{GF}^{(2)}(\vec{\theta}_{2},\vec{\phi}_{2})
+        \hat{U}_{GF}^{(1)}(\vec{\theta}_{1},\vec{\phi}_{1}) \vert HF \rangle
+
+    where each of the gate fabric block :math:`\hat{U}_{GF}(\vec{\theta},\vec{\phi})` comprises of 2-parameter 4-qubit
+    gate elements :math:`\hat{Q}(\theta, \phi)` that act on 4-nearest-neighboring qubits. The circuit implementing a
+    single layer of the gate fabric block for :math:`N = 8` is shown in the figure below:
+
+    .. figure:: ../../_static/templates/layers/gate_fabric_layer.png
+        :align: center
+        :width: 80%
+        :target: javascript:void(0);
+
+    The gate element :math:`\hat{Q}(\theta, \phi)` (`arXiv:2104.05692 <https://arxiv.org/abs/2104.05695>`_) is composed of
+    a 1-parameter 4-qubit spin-adapted spatial orbital rotation gate, which is implemented by :class:`~.OrbitalRotation()`
+    operation and 1-parameter 4-qubit diagonal pair-exchange gate, which is equivalent to the :class:`~.DoubleExcitation()`
+    operation. In addition to these two gates, the gate element :math:`\hat{Q}(\theta, \phi)` can also include an optional
+    constant :math:`\hat{\Pi} \in \{\hat{I}, \text{OrbitalRotation}(\pi)\}` gate.
+
+    .. figure:: ../../_static/templates/layers/q_gate_decompositon.png
+        :align: center
+        :width: 75%
+        :target: javascript:void(0);
+
+    |
+
+    The individual four qubit :class:`~.DoubleExcitation()` and :class:`~.OrbitalRotation()` gates given here are equivalent to the
+    :math:`\text{QNP}_{PX}(\theta)` and :math:`\text{QNP}_{OR}(\phi)` gates presented in `arXiv:2104.05692 <https://arxiv.org/abs/2104.05695>`_
+    respectively. Moreover, regardless of the choice of :math:`\hat{\Pi}`, this gate fabric will exactly preserve the quantum numbers
+    :math:`\hat{N}_{\alpha}`, :math:`\hat{N}_{\beta}` and :math:`\hat{S}^{2}`.
+
+    Args:
+        weights (tensor_like): Array of weights of shape ``(D, L, 2)``\,
+            where ``D`` is the number of gate fabric layers and ``L = N/2-1``
```

Comment:
What is `N` here? The number of qubits? Good to specify again even if we did earlier on in the docstring